### ISSUE TITLE
Should I provide the schemas when creating a prediction engine?

### ISSUE THREAD

**Issue by PeterPann23** [dotnet/machinelearning#3378]
[Enter feedback here]
currently, it takes quite a while to generate a prediction engine (56,863 ms). 
Would serialize and providing the schemas generated by the training improve the load times?

A single prediction takes 1,452 ms. I have an I7 CPU with sufficient ram, is there a way to speed it up?

I have added the input class. 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 992ff2a9-37e1-54ce-947b-67103207ce8e
* Version Independent ID: 584449fa-50f6-85af-4c4c-3fb33cdedfbd
* Content: [ModelOperationsCatalog.CreatePredictionEngine Method (Microsoft.ML)](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.modeloperationscatalog.createpredictionengine?view=ml-dotnet)
* Content Source: [dotnet/xml/Microsoft.ML/ModelOperationsCatalog.xml](https://github.com/dotnet/ml-api-docs/blob/live/dotnet/xml/Microsoft.ML/ModelOperationsCatalog.xml)
* Product: **dotnet-ml-api**
* GitHub Login: @sfilipi
* Microsoft Alias: **johalex**

**Comment by glebuk** [comment]
@PeterPann23,
Excellent question, I believe we ran into this issue ourselves recently. The following PR should address it #3185.  Can you see if this fixes it for you?  If not, can you describe the pipeline you are trying to save?  It should take few milliseconds.

**Comment by PeterPann23** [comment]
@gelb k
I god a weird error stating it as missing some field that I never hear of, looked it up and it was used as a default value to group on.

If I do not provide it (LightGmb) then it works like a charm… slow as hell but one I have the predictor I can keep it in memory. The first prediction is also really slow, the consecutive prediction are again fast <7ms on my pc. I have a use case with +35k properties and I score >98.5% on my test set, I am now building a production set so not sure what my score is.

I do not have much faith in the score provided by the metrics on the test dataset, it seems to be really off to that what I get when I run it using the prediction engine against a poco class.



Gesendet von Mail<https://go.microsoft.com/fwlink/?LinkId=550986> für Windows 10

________________________________
Von: Gleb K <notifications@github.com>
Gesendet: Friday, April 19, 2019 8:03:11 AM
An: dotnet/machinelearning
Cc: PeterPann23; Mention
Betreff: Re: [dotnet/machinelearning] Should I provide the schemas when creating a prediction engine? (#3378)


@PeterPann23<https://github.com/PeterPann23>,
Excellent question, I believe we ran into this issue ourselves recently. The following PR should address it #3185<https://github.com/dotnet/machinelearning/pull/3185>. Can you see if this fixes it for you? If not, can you describe the pipeline you are trying to save? It should take few milliseconds.

**Comment by glebuk** [comment]
This is a bug that was fixed after the 1.0.pre has been released.  Can you try to build from master to see if the latest bits fix your issue? ?  t should accelerate prediction engine creation 100x -1000x.
Gleb

**Comment by PeterPann23** [comment]
So I did that, one Vector, and binary file… no “Speed improvement” till now.  As it is still running after 40+ hours, that’s till now 1.5x slower and it’s still going




Gesendet von Mail<https://go.microsoft.com/fwlink/?LinkId=550986> für Windows 10

________________________________
Von: Gleb K <notifications@github.com>
Gesendet: Monday, April 22, 2019 3:39:11 AM
An: dotnet/machinelearning
Cc: PeterPann23; Mention
Betreff: Re: [dotnet/machinelearning] Should I provide the schemas when creating a prediction engine? (#3378)


This is a bug that was fixed after the 1.0.pre has been released. Can you try to build from master to see if the latest bits fix your issue? ? t should accelerate prediction engine creation 100x -1000x.
Gleb

**Comment by glebuk** [comment]
@PeterPann23,, the issue still does not seem to be resoved for you. have you tried other suggestions I've made in other issues? Specifically, are you reading the data as 23k columns or a few feature vectors? Such slow loading is not at all expected. It should take single milliseconds.

**Comment by PeterPann23** [comment]
I have tried, slowest-> fastest:
- Combine the columns concatenate to features vector in pipeline from a csv to binary converted
- Combine the columns concatenate to features vector in pipeline
- Load as a vector from CSV
- Load as a vector from a csv to binary converted

I was surprised that the single column, converted to binary was actually the slowest. 
I have several training sets, the ones I tried have 35,630 and 40,730 feature columns, not sure what the 23K you mention is.

the 40,730 feature columns sample file size for the  csv is 609,702,654 KB (ASCII encoded) , the same file in binary format is 58,691,090 KB and contain 14 days of data. All systems tested on the same physical hardware
